### PR TITLE
New syscalls

### PR DESCRIPTION
**Pull request by osipov-mit**
- Add new syscalls to env to instantiate wasm module
- Split getwasmmetadata and reading state to separate functions
- Update readState function
- Bump @polkadot/api to `7.10.1`
- Bump version to `0.14.2`

### CODE REVIEW

**Review by EugenWay — APPROVED** [review]

File: api/src/wasm/readState.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import importObj from './importObj';
+import { getExportValue } from './utils';
+
+export async function readState(wasmBytes: Buffer, pages: any, inputValue?: Uint8Array): Promise<Uint8Array> {
+  const memory = new WebAssembly.Memory({ initial: Object.keys(pages).length });
+  let module = await WebAssembly.instantiate(wasmBytes, importObj(memory, false, inputValue));
```
Comment (nikitayutanov) [inline api/src/wasm/readState.ts:6]:
```suggestion
  const module = await WebAssembly.instantiate(wasmBytes, importObj(memory, false, inputValue));
```

File: api/src/wasm/readState.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  Object.keys(pages).forEach((pageNumber: string) => {
+    const start = +pageNumber * 65536;
+    const end = start + 65536;
+    const page = pages[pageNumber];
+    for (let i = start; i < end; i++) {
+      new Uint8Array(memory.buffer)[i] = page[i % 65536];
+    }
+  });
+  const exports = module.instance.exports;
```
Comment (nikitayutanov) [inline api/src/wasm/readState.ts:15]:
```suggestion
  const { exports } = module.instance;
```

File: api/src/wasm/readState.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  if (!exports) {
+    return null;
+  }
+  return new Uint8Array(getExportValue(memory, exports.meta_state));
```
Comment (nikitayutanov) [inline api/src/wasm/readState.ts:19]:
```suggestion
  return exports ? new Uint8Array(getExportValue(memory, exports.meta_state)) : null;
```

File: api/src/wasm/utils.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+}
+
+export function getExportValue(memory: WebAssembly.Memory, func: any): ArrayBuffer {
+  if (!func) {
+    return undefined;
+  }
+  let result_ptr = func();
+  let pointer = new Uint32Array(memory.buffer.slice(result_ptr, result_ptr + 4))[0];
+  let length = new Uint32Array(memory.buffer.slice(result_ptr + 4, result_ptr + 8))[0];
+  let buf = memory.buffer.slice(pointer, pointer + length);
```
Comment (nikitayutanov) [inline api/src/wasm/utils.ts:12]: `const` instead of `let`?

File: api/src/wasm/utils.ts [hunk 5]
```diff
@@ -16,3 +18,17 @@ export function getExportValue(memory: WebAssembly.Memory, func: any): ArrayBuff
 export function ab2str(buf: ArrayBuffer): string {
   return String.fromCharCode.apply(null, new Uint8Array(buf));
 }
+
+export function getInitialLength(pages: IGearPages) {
+  return (
+    parseInt(
+      Object.keys(pages).reduce((prev, cur) => {
+        console.log({ prev, cur });
```
Comment (nikitayutanov) [inline api/src/wasm/utils.ts:26]: Should it remain?

File: api/src/State.ts [hunk 6]
```diff
@@ -47,10 +43,20 @@ export class GearProgramState extends GearStorage {
...
+      throw new ReadStateError(`Unable to read state. meta_state_output type is not specified in metadata`);
+    }
+    if (metadata.meta_state_input && inputValue === undefined) {
+      throw new ReadStateError(`Unable to read state. inputValue not specified`);
+    }
+    const encodedInput = inputValue === undefined ? undefined : await this.encodeInput(metadata, inputValue);
+    const state = await readState(metaWasm, pages, encodedInput);
 
-    return await this.decodeState(metaWasm, pages, encodedInput);
+    return await this.decodeState(state, metadata);
```
Comment (nikitayutanov) [inline api/src/State.ts:60]:
```suggestion
    return this.decodeState(state, metadata);
```
[`return await` is redudant](https://eslint.org/docs/rules/no-return-await)

File: api/src/wasm/readState.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  const memory = new WebAssembly.Memory({ initial: getInitialLength(pages) });
+  const module = await WebAssembly.instantiate(wasmBytes, importObj(memory, false, inputValue));
+  Object.keys(pages).forEach((pageNumber: string) => {
+    const start = +pageNumber * 65536;
+    const end = start + 65536;
+    const page = pages[pageNumber];
+    for (let i = start; i < end; i++) {
+      new Uint8Array(memory.buffer)[i] = page[i % 65536];
+    }
+  });
```
Comment (nikitayutanov) [inline api/src/wasm/readState.ts:15]: Define `65536` as a variable.
Also for me it's not clear what `i % 65536` represents, so I believe it's better to declare and name it too.

**Review by aderrod94 — APPROVED** [review]

**Review by nikitayutanov — APPROVED** [review]

**Pull request merged**